clang/include/clang/Tooling/DependencyScanning/DependencyScanningFilesystem.h

```diff
@@ -230,6 +251,26 @@ class DependencyScanningFilesystemLocalCache {
     assert(InsertedEntry == &Entry && "entry already present");
     return *InsertedEntry;
   }
+
+  /// Returns real path associated with the filename or nullptr if none is
+  /// found.
+  const CachedRealPath *findRealPathByFilename(StringRef Filename) const {
+    assert(llvm::sys::path::is_absolute_gnu(Filename));
+    auto It = RealPathCache.find(Filename);
+    return It == RealPathCache.end() ? nullptr : It->getValue();
+  }
+
+  /// Associates the given real path with the filename and returns the given
+  /// entry pointer (for convenience).
+  const CachedRealPath &
+  insertRealPathForFilename(StringRef Filename,
+                            const CachedRealPath &RealPath) {
```

Comment:
Since you are storing a pointer to this not the value I would suggest this take a pointer parameter. Otherwise it's easy to assume it's being copied and would be safe to pass a stack variable.

Reply:
I see your point. I used a reference here to be consistent with existing code in this file. I think the original motivation for using references was the "non-null pointer" aspect of it. I tried converting all uses of `const CachedFileSystemEntry &` in the existing code to pointers, but that's impossible to understand. Doing that just for arguments to insert-like functions is a bit better, but I'm not sure it's clear win in readability/understandability.

Reply:
Okay, I'm okay with being consistent for now, but I think it's a bad pattern in general to pass around `const X &` and keep a pointer to it.

Reply:
Agreed. I'll go over this file and probably turn all nullable pointers into `std::optional<const X *>` (AFAIK these are only function return types and don't get stored in long-lived data structures) and assume plain `const X *` are not null.